### ISSUE TITLE
segmentation fault while reproducing 'R50-C4'

### ISSUE THREAD

**Issue by sehyun03** [tensorpack/tensorpack#861]
Thanks for great work.

I used 2 GPU with horovod.
While training example 'R50-C4' segmentation fault occurs. 
Also fastrcnn-losses show 'nan' (not sure its relevant to crashing issues)

An issue :
- Unexpected Problems / Potential Bugs

1. What you did:
  + If you're using examples:
    + What's the command you run:
       

> CUDA_VISIBLE_DEVICES=0,1 mpirun -np 2 --output-filename R50_pretrained_log python3 ./train.py --logdir=./train_log/Fasternn_R50_ImageNet_pretrained --config MODE_MASK=False FRCNN.BATCH_PER_IM=64 PREPROC.SHORT_EDGE_SIZE=600 PREPROC.MAX_SIZE=1024 TRAIN

.LR_SCHEDULE=[100000,230000,280000] TRAIN.NUM_GPUS=2

    + Have you made any changes to code? Paste them if any:

On "config.py"

> _C.TRAINER = 'horovod'
> _C.DATA.BASEDIR = '/home/cvhsh/proj_knowledge_trans_tspack/COCO'
> _C.BACKBONE.WEIGHTS = '/home/cvhsh/proj_knowledge_trans_tspack/COCO/ImageNet-ResNet50.npz'   # /path/to/weights.npz

2. What you observed, including but not limited to the __entire__ logs.
  + Better to paste what you observed instead of describing them.

> [0813 19:28:17 @base.py:237] Start Epoch 52 ...
> 
> 100%|##########|500/500[01:47<00:00, 4.65it/s][0813 19:30:04 @base.py:247] Epoch 52 (global_step 26000) finished, time:1 minute 47 seconds.
> [0813 19:30:04 @graph.py:73] Running Op horovod_broadcast/group_deps ...
> 100%|##########|500/500[01:47<00:00, 4.64it/s][0813 19:30:04 @base.py:247] Epoch 52 (global_step 26000) finished, time:1 minute 47 seconds.
> [0813 19:30:04 @graph.py:73] Running Op horovod_broadcast/group_deps ...
> [0813 19:30:05 @monitor.py:435] PeakMemory(MB)/gpu:0: 4393.2
> [0813 19:30:05 @base.py:237] Start Epoch 53 ...
> [0813 19:30:05 @misc.py:99] Estimated Time Left: 2 days 17 hours 40 minutes 46 seconds
> 
>   0%|          |0/500[00:00<?,?it/s][0813 19:30:05 @monitor.py:435] PeakMemory(MB)/gpu:0: 4418.7
> [0813 19:30:05 @monitor.py:435] QueueInput/queue_size: 50
> [0813 19:30:05 @monitor.py:435] fastrcnn_losses/box_loss: nan
> [0813 19:30:05 @monitor.py:435] fastrcnn_losses/label_loss: nan
> [0813 19:30:05 @monitor.py:435] fastrcnn_losses/label_metrics/accuracy: 0.76824
> [0813 19:30:05 @monitor.py:435] fastrcnn_losses/label_metrics/false_negative: 1
> [0813 19:30:05 @monitor.py:435] fastrcnn_losses/label_metrics/fg_accuracy: 2.246e-37
> [0813 19:30:05 @monitor.py:435] learning_rate: 0.01
> [0813 19:30:05 @monitor.py:435] rpn_losses/box_loss: 0.056564
> [0813 19:30:05 @monitor.py:435] rpn_losses/label_loss: 0.24104
> [0813 19:30:05 @monitor.py:435] rpn_losses/label_metrics/precision_th0.1: 0.38716
> [0813 19:30:05 @monitor.py:435] rpn_losses/label_metrics/precision_th0.2: 0.4666
> [0813 19:30:05 @monitor.py:435] rpn_losses/label_metrics/precision_th0.5: 0.54889
> [0813 19:30:05 @monitor.py:435] rpn_losses/label_metrics/recall_th0.1: 0.79813
> [0813 19:30:05 @monitor.py:435] rpn_losses/label_metrics/recall_th0.2: 0.71973
> [0813 19:30:05 @monitor.py:435] rpn_losses/label_metrics/recall_th0.5: 0.3714
> [0813 19:30:05 @monitor.py:435] rpn_losses/num_pos_anchor: 40.375
> [0813 19:30:05 @monitor.py:435] rpn_losses/num_valid_anchor: 256
> [0813 19:30:05 @monitor.py:435] sample_fast_rcnn_targets/num_bg: 49.167
> [0813 19:30:05 @monitor.py:435] sample_fast_rcnn_targets/num_fg: 14.833
> [0813 19:30:05 @monitor.py:435] sample_fast_rcnn_targets/proposal_metrics/best_iou_per_gt: nan
> [0813 19:30:05 @monitor.py:435] sample_fast_rcnn_targets/proposal_metrics/recall_iou0.3: 0.63788
> [0813 19:30:05 @monitor.py:435] sample_fast_rcnn_targets/proposal_metrics/recall_iou0.5: 0.53133
> [0813 19:30:05 @monitor.py:435] total_cost: nan
> [0813 19:30:05 @monitor.py:435] wd_cost: nan
> [0813 19:30:05 @base.py:237] Start Epoch 53 ...
> 
>   0%|          |0/500[00:00<?,?it/s][cvhsh-labdevice:25111] *** Process received signal ***
> [cvhsh-labdevice:25111] Signal: Segmentation fault (11)
> [cvhsh-labdevice:25111] Signal code: Address not mapped (1)
> [cvhsh-labdevice:25111] Failing at address: (nil)
> [cvhsh-labdevice:25111] [ 0] /lib/x86_64-linux-gnu/libpthread.so.0(+0x11390)[0x7f7ddd406390]
> [cvhsh-labdevice:25111] [ 1] /usr/local/lib/python3.5/dist-packages/horovod/common/mpi_lib.cpython-35m-x86_64-linux-gnu.so(+0x22047)[0x7f7d2cf2f047]
> [cvhsh-labdevice:25111] [ 2] /usr/local/lib/python3.5/dist-packages/horovod/common/mpi_lib.cpython-35m-x86_64-linux-gnu.so(+0x2409b)[0x7f7d2cf3109b]
> [cvhsh-labdevice:25111] [ 3] /usr/lib/x86_64-linux-gnu/libstdc++.so.6(+0xb8c80)[0x7f7dd5c0fc80]
> [cvhsh-labdevice:25111] [ 4] /lib/x86_64-linux-gnu/libpthread.so.0(+0x76ba)[0x7f7ddd3fc6ba]
> [cvhsh-labdevice:25111] [ 5] /lib/x86_64-linux-gnu/libc.so.6(clone+0x6d)[0x7f7ddd13241d]
> [cvhsh-labdevice:25111] *** End of error message ***
> \-------------------------------------------------------
> Primary job  terminated normally, but 1 process returned
> a non-zero exit code. Per user-direction, the job has been aborted.
> \-------------------------------------------------------
> \--------------------------------------------------------------------------
> mpirun noticed that process rank 0 with PID 0 on node cvhsh-labdevice exited on signal 11 (Segmentation fault).
> \--------------------------------------------------------------------------

3. What you expected, if not obvious.
4. Your environment:
  + Python version.
     : 3.5.2
  + TF version: `python -c 'import tensorflow as tf; print(tf.GIT_VERSION, tf.VERSION)'`.
     : 1.9.0
  + Tensorpack version: `python -c 'import tensorpack; print(tensorpack.__version__)'`.
     : 0.8.6
  + Hardware information, if relevant.
      : GPU PASCAL TITAN Xp 12G
  + others
      : ubuntu 16.04, nvidia drivier 396.37, CUDA 9.0
      : horovod(0.13.11),  mpirun (Open MPI) 3.1.1, opencv-contrib-python (3.4.2.17)

**Comment by ppwwyyxx** [comment]
I've run into the same issue (not NaN, but segfault) with horovod before, not sure which library is causing it.
The replicated trainer should be more stable to use.

**Comment by ppwwyyxx** [comment]
It may also help to set `NCCL_DEBUG=INFO` to see if there is any issue in communication. Also make sure horovod was rebuilt and reinstalled after each time you install tensorflow.

**Comment by ppwwyyxx** [comment]
The NaN issue turns out to because the learning rate wasn't correctly scaled for <8 GPUs. The code was only tested with 8+ GPUs. I'll push a fix for that one later.

**Comment by ppwwyyxx** [comment]
Closing as the crash does not seem to be a tensorpack issue. Uninstall and reinstall horovod is worth trying.
